plugins/inputs/intel_powerstat/intel_powerstat.go

```diff
@@ -229,6 +268,153 @@ func (p *PowerStat) addCPUTemperatureMetric(cpuID string, acc telegraf.Accumulat
 	acc.AddGauge("powerstat_core", fields, tags)
 }
 
+func calculateTurboRatioGroup(coreCounts uint64, msr uint64, group map[int]uint64) {
+	from := coreCounts & 0xFF // value of number of active cores of bucket 1 is written in the first 8 bits. The next buckets values are saved on the following 8-bit sides
+	for i := 0; i < 8; i++ {
+		to := (coreCounts >> (i * 8)) & 0xFF
+		if to == 0 {
+			break
+		}
+		value := (msr >> (i * 8)) & 0xFF
+		// value of freq ratio is stored in 8-bit blocks, and their real value is obtained after multiplication by 100
+		if value != 0 && to != 0 {
+			for ; from <= to; from++ {
+				group[int(from)] = value * 100
+			}
+		}
+		from = to + 1
+	}
+}
+
+func (p *PowerStat) addTurboRatioLimit(socketID string, acc telegraf.Accumulator) {
```

Comment:
Different CPU models record their turbo ratio limit values in a different way and/or in a different MSR. The method that has been implemented supports not only older but also latest CPUs (eg Sapphire Rapids), and their specific metric delivery methods. 
The operation logic is similar to function ``` static void dump_cstate_pstate_config_info (unsigned int family, unsigned int model)``` in ```turbostat``` https://github.com/torvalds/linux/blob/v5.18-rc6/tools/power/x86/turbostat/turbostat.c#L3975
If there are new processor models with a different way of specifying turbo ratio limit metrics, it will be necessary to include it in the above code.